src/main/java/net/fabricmc/tinyremapper/util/VersionAwareMap.java

```diff
@@ -0,0 +1,373 @@
+     *                                       is not supported by this map
+     * @throws ClassCastException            if the class of the specified key or value
+     *                                       prevents it from being stored in this map
+     * @throws NullPointerException          if the specified key or value is null
+     *                                       and this map does not permit null keys or values
+     * @throws IllegalArgumentException      if some property of the specified key
+     *                                       or value prevents it from being stored in this map
+     */
+    @Override
+    public ClassInstance put(VersionedName key, ClassInstance value) {
+        ClassInstance result = remove(key);
+        storage.get(key.getName()).add(value);
+        return result;
+    }
+
+    /**
+     * Removes the mapping for a key from this map if it is present
+     * (optional operation).   More formally, if this map contains a mapping
+     * from key {@code k} to value {@code v} such that
+     * {@code Objects.equals(key, k)}, that mapping
+     * is removed.  (The map can contain at most one such mapping.)
+     *
+     * <p>Returns the value to which this map previously associated the key,
+     * or {@code null} if the map contained no mapping for the key.
+     *
+     * <p>If this map permits null values, then a return value of
+     * {@code null} does not <i>necessarily</i> indicate that the map
+     * contained no mapping for the key; it's also possible that the map
+     * explicitly mapped the key to {@code null}.
+     *
+     * <p>The map will not contain a mapping for the specified key once the
+     * call returns.
+     *
+     * @param key key whose mapping is to be removed from the map
+     * @return the previous value associated with {@code key}, or
+     * {@code null} if there was no mapping for {@code key}.
+     * @throws UnsupportedOperationException if the {@code remove} operation
+     *                                       is not supported by this map
+     * @throws ClassCastException            if the key is of an inappropriate type for
+     *                                       this map
+     *                                       (<a href="{@docRoot}/java.base/java/util/Collection.html#optional-restrictions">optional</a>)
+     * @throws NullPointerException          if the specified key is null and this
+     *                                       map does not permit null keys
+     *                                       (<a href="{@docRoot}/java.base/java/util/Collection.html#optional-restrictions">optional</a>)
+     */
+    @Override
+    public ClassInstance remove(Object key) {
+        ClassInstance result = get(key);
+
+        if (key instanceof VersionedName) {
+            String name = ((VersionedName) key).getName();
+            storage.get(name).removeIf(cls -> cls.getVersionedName().equals(key));
```

Comment:
Instead of calling `get`, you should just do:
```java
ClassInstance[] removed = new ClassInstance[1];
storage.get(name).removeIf(cls -> {
  if (cls.getVersionedName().equals(key)) {
    removed[0] = cls;
    return true;
  }
  return false;
});
return removed[0];
```
this is better in terms of performance.

Reply:
I agree. I will do this performance patch after everything is working.

Reply:
Or you can use a for loop to find and remove the object. for loop would actually be faster because we are sure there is only one matching element and we can break the loop early once we found it